concurrency: emit structured contention information to trace

tbg opened the pull request:
This change attaches a protobuf payload to the current Span whenever a
request conflicts with another transaction. The payload contains the
contending txn (i.e. the pushee) at the time at which it was first
encountered, the key on which the conflict occurred (note that this is
not necessarily the key at which the pushee is anchored) and the time
spent waiting on the conflict (excluding intent resolution).

This enables #57114.

Touches #55583. I am not closing that issue yet because we also want
to potentially track the outcome of the conflict.

Release note: None

tbg committed:
concurrency: emit structured contention information to trace

This change attaches a protobuf payload to the current Span whenever a
request conflicts with another transaction. The payload contains the
contending txn (i.e. the pushee) at the time at which it was first
encountered, the key on which the conflict occurred (note that this is
not necessarily the key at which the pushee is anchored) and the time
spent waiting on the conflict (excluding intent resolution).

This enables #57114.

Touches #55583. I am not closing that issue yet because we also want
to potentially track the outcome of the conflict.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58444)
<!-- Reviewable:end -->

nvb reviewed (commented):
> is this basic approach sane or should contention metadata be generated elsewhere (i.e. not in `WaitOn`)?

Yes, `WaitOn` seems like the right place for this. The `lockTableWaiter` is where all of the waiting on locks is coordinated.

> In particular, it's unclear to me whether `state.txn` and/or `state.key` can change within a single invocation of `WaitOn`.

Both of these fields can change while iterating in `WaitOn`. The idea there is that we grab a snapshot of the lockTable, iterate over it, and wait on each lock that's in our way. `WaitOn` continues to listen to these updates on its `newStateC`, clearing the way if necessary by pushing, until it eventually reaches a `doneWaiting` state.

> Completing the current approach would likely result in spaghetti code that is easy to get wrong. We seem to want to keep a bit of state around; more than we do with the current structure which intentionally minimizes state held across iterations of the wait loop.

Yeah, I think what you have here is a little different than how I would do it. I was imagining that you'd begin a new span when you receive on `newStateC` (with kind `waitFor`/`waitForDistinguished`) and then close that span when you receive the next update on `newStateC` (with any kind). So each span's lifetime is the time between updates on `newStateC`.

> Reviewed 2 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-:-MQIaORM6v5ivswkAWyP:bs0c7f9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQIai-Q8IyamDaABVWr:-MQIai-Q8IyamDaABVWs:b-b22xoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/aedaab4804042612e7b636c3998bb1529f0a7c7d/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L165)):*
> ```Go
> 			switch state.kind {
> 			case waitFor, waitForDistinguished:
> 				// NB: TxnLockTableWaiterWithNonTxn has state.{key,txn}==nil, despite
> ```

Yeah, we should fix the test instead of adding this conditional. It really doesn't make sense for either of these fields to be `nil`.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Both of these fields can change while iterating in WaitOn. The idea there is that we grab a snapshot of the lockTable, iterate over it, and wait on each lock that's in our way. WaitOn continues to listen to these updates on its newStateC, clearing the way if necessary by pushing, until it eventually reaches a doneWaiting state.

Right, but can something like this happen:

- waitFor txn1 on key1, decide to do something (maybe we do a liveness push and txn is still live)
- go back to waiting
- waitFor gives us txn2 on key 2, decide to do something, go back to waiting
- now txn1 on key1 comes back
- ...

I want to avoid emitting a payload for each time we see a given txn; that could be an unbounded amount, no? Plus these individual updates don't mean much to the caller - the aggregate does (maybe that's wrong; not sure). Do I need to maintain a map in this method?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-:-MQNHX7947noZDqS4Nzz:b-wd78ue)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
We chatted directly, updated the PR. Still some cleanup remaining but I think a review pass is reasonable. Note that when I rebased and picked up recent PRs in this area, a got a few diffs:

```
diff --git a/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/clear_abandoned_intents b/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/clear_abandoned_intents
index 4fdef227d6..8f3468882d 100644
--- a/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/clear_abandoned_intents
+++ b/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/clear_abandoned_intents
@@ -116,15 +116,6 @@ on-txn-updated txn=txn2 status=aborted
 [3] sequence req1: scanning lock table for conflicting locks
 [3] sequence req1: sequencing complete, returned guard
 [3] sequence req1: contention metadata: 00000002 @ "a"
-[3] sequence req1: contention metadata: 00000002 @ "b"
-[3] sequence req1: contention metadata: 00000002 @ "c"
-[3] sequence req1: contention metadata: 00000002 @ "d"
-[3] sequence req1: contention metadata: 00000002 @ "e"
-[3] sequence req1: contention metadata: 00000002 @ "f"
-[3] sequence req1: contention metadata: 00000002 @ "g"
-[3] sequence req1: contention metadata: 00000002 @ "h"
-[3] sequence req1: contention metadata: 00000002 @ "i"
-[3] sequence req1: contention metadata: 00000002 @ "j"
```

I assume they are a consequence of moving the txn status cache query and intent resolution above `WaitOn`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-:-MQNeWZhAUFGZWyIWFPS:bpty8ej)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg committed:
tracing: offer structured payloads to verbose log

When a payload is added, and the Span is verbose, also
stringify the payload into the verbose log.

Release note: None

tbg commented:
cc @irfansharif and @knz for pod visibility.

@nvanbenschoten this is now ready for review.

nvb reviewed (commented):
Nice and clean!

> Reviewed 1 of 14 files at r2, 17 of 21 files at r3, 2 of 5 files at r4, 1 of 1 files at r7, 21 of 21 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-:-MQVPIrjD7HWhc1HxCiE:btraywv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager_test.go, line 914 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVONsJ4YVRqa_GUW_q:-MQVONsKD89l5cw--qBh:btszc48) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/concurrency_manager_test.go#L914)):*
> ```Go
> 		rec := g.collect()
> 		for _, span := range rec {
> 			for _, log := range span.Logs {
> ```

Out of curiosity, why did you need to pull out the structured events separately in an earlier revision of this PR, but no longer need to do so? 
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 128 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVLbz_6_2rzmXxt3q0:-MQVLbz_6_2rzmXxt3q1:bucx9ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L128)):*
> ```Go
> }
> 
> // contentionEventHelper tracks and emits ContentionEvents.
> ```

As a matter of style, I think we should move this helper and its method down below the `lockTableWaiterImpl` methods, maybe right after `txnCache` and its methods. Most readers of this file will not need to know about this `contentionEventHelper`, so this is a lot of noise right up top.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 163 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVOqn6E-gDJzJycNr-:-MQVOqn6E-gDJzJycNr0:b6xq9ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L163)):*
> ```Go
> 		ice.sp.LogStructured(ice.ev)
> 	}
> 	ice.ev = nil
> ```

nit: move this one line up so that this is more easily read as "if an event already exists, take it and process it".
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 186 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVMcij4OtTLnxKjDgz:-MQVMcij4OtTLnxKjDh-:b8z7245) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L186)):*
> ```Go
> 	var timerWaitingState waitingState
> 
> 	ice := contentionEventHelper{
> ```

Why `ice`? What does that stand for?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 204 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVLQc29mCZVBvE7N9X:-MQVLQc29mCZVBvE7N9Y:b-oc61mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L204)):*
> ```Go
> 			switch state.kind {
> 			case waitFor, waitForDistinguished:
> 				ice.emitAndInit(state.key, *state.txn)
> ```

I think we're going to need to handle other `waitKind`s as well, not just `waitFor` and `waitForDistinguished`. While the other `waitKind`s might not create contention events, they'll certainly want to finalize the existing event.

This also makes it easier to extend this in the future to create contention events for other `waitKind`s. For instance, `waitSelf` may be an eventual candidate, though it should be super rare, so I wouldn't both doing anything special for it now.

So that indicates that we may just want to hand `contentionEventHelper` the entire `waitingState` (after moving this up two lines) and let it decide what to do for each `waitKind`.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! This should be ready to go now.

 I added a commit to handle waitSelf properly. I would like to exercise this through `TestConcurrencyManager`, but had some trouble (since I also don't exactly understand how to trigger the waitSelf state through its primitives). Help with that is appricated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-:-MQkv3mw4oZa0PTqdTKT:bejxrdx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager_test.go, line 914 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVONsJ4YVRqa_GUW_q:-MQkmrnM017bY2XZBRzS:b-mwq7k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/concurrency_manager_test.go#L914)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Out of curiosity, why did you need to pull out the structured events separately in an earlier revision of this PR, but no longer need to do so? 
</blockquote></details>

Because I wanted to print them, but I changed things so that on a verbose span, payloads are stringified into the log so they're already here (and in the right order relative to regular log messages).
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 186 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVMcij4OtTLnxKjDgz:-MQknQv39OdQBLHxLoSW:bpyt4n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L186)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why `ice`? What does that stand for?
</blockquote></details>

It abbreviated some earlier name for `contentionEventHelper`. Fixed.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 204 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVLQc29mCZVBvE7N9X:-MQkuzH53NRQyBs7VBa4:b-6z7wvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L204)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think we're going to need to handle other `waitKind`s as well, not just `waitFor` and `waitForDistinguished`. While the other `waitKind`s might not create contention events, they'll certainly want to finalize the existing event.

This also makes it easier to extend this in the future to create contention events for other `waitKind`s. For instance, `waitSelf` may be an eventual candidate, though it should be super rare, so I wouldn't both doing anything special for it now.

So that indicates that we may just want to hand `contentionEventHelper` the entire `waitingState` (after moving this up two lines) and let it decide what to do for each `waitKind`.
</blockquote></details>

Good idea - done. I stayed away from waitSelf but I think I want to fix it in a follow-up.


<!-- Sent from Reviewable.io -->

tbg commented:
@nvanbenschoten PTAL at the waitSelf commit. I tried writing the concurrency manager test that exercises waitSelf. I don't think I managed, but it is hopefully close. The test file has comments explaining where I'm stuck.

tbg commented on pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self:
```diff
@@ -0,0 +1,109 @@
+[2] sequence reqTxn1: acquiring latches
+[2] sequence reqTxn1: blocked on select in spanlatch.(*Manager).waitForSignal
+
+sequence req=reqTxnMiddle
```
move this to later to avoid any chance of racing with reqTxn1

tbg commented:
This is ready for a look, @nvanbenschoten.

irfansharif commented on pkg/roachpb/api.proto:
```diff
@@ -2203,14 +2203,13 @@ service Internal {
+
   // Key is the key that this and the other transaction conflicted on.
   bytes key = 1 [(gogoproto.casttype) = "Key"];
   // Txn is the other transaction.
```
s/Txn/TxnMeta

irfansharif commented on pkg/kv/kvserver/concurrency/lock_table_waiter.go:
```diff
@@ -648,6 +666,68 @@ func (c *txnCache) insertFrontLocked(txn *roachpb.Transaction) {
+
+// emitAndInit compares the waitingState's active txn (if any) against the open
+// ContentionEvent (if any). If the they match, we are continuing to handle the
+// same event and no action is taken. If they differ, the open event (if any) is
```
It would be a bit easier to read the implementation below if we had an explicit early return for this case.

irfansharif commented on pkg/kv/kvserver/concurrency/lock_table_waiter.go:
```diff
@@ -648,6 +666,68 @@ func (c *txnCache) insertFrontLocked(txn *roachpb.Transaction) {
+	h.ev = nil
+}
+
+// emitAndInit compares the waitingState's active txn (if any) against the open
```
s/open/{last, existing}?

irfansharif commented on pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self:
```diff
@@ -0,0 +1,130 @@
+#
+# TODO(nvanbenschoten): would be nice to make this more explicit, but I'm not sure
+# how to change `(*lockTableImpl).Format` to make it happen in debug-lock-table above.
+# doesn't
```
Stray "doesn't"

irfansharif reviewed (approved):
LGTM, I'll let Nathan sign off on waitSelf test.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 27 files at r9, 22 of 22 files at r10, 2 of 2 files at r11, 1 of 24 files at r13, 19 of 22 files at r14, 4 of 4 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-:-MRR4_pD1A7JeX2pFlsv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager_test.go, line 213 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MRR3Dsc1_t-hFmjLz66:-MRR3Dsc1_t-hFmjLz67:b-3dmbcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/756037b892f1d2ef18b74b2435459ff76b844b47/pkg/kv/kvserver/concurrency/concurrency_manager_test.go#L213)):*
> ```Go
> 				return c.waitAndCollect(t, mon)
> 
> 			case "cancel":
> ```

We don't need this anymore, right? Were you intending to remove it?
___
*[pkg/kv/kvserver/concurrency/concurrency_manager_test.go, line 709 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MRR3HWn7KgvRHkxEgt2:-MRR3HWn7KgvRHkxEgt3:b-b2jx3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/756037b892f1d2ef18b74b2435459ff76b844b47/pkg/kv/kvserver/concurrency/concurrency_manager_test.go#L709)):*
> ```Go
> 	defer c.mu.Unlock()
> 	if p, ok := c.txnPushes[pusher]; ok {
> 		// TODO(tbg): assert same pushee
> ```

Can we remove the TODO?
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 83 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQlnt0ECz6Yv9Sp91K6:-MQlnt0ECz6Yv9Sp91K7:b98zuee) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc22165dc0c549ddd0586ec83c904bfa66187d44/pkg/kv/kvserver/concurrency/lock_table.go#L83)):*
> ```Go
> 	// transaction may be a lock holder of a conflicting lock or a
> 	// conflicting request being sequenced through the same lockTable.
> 	txn  *enginepb.TxnMeta // always non-nil in waitFor{,Distinguished}
> ```

`,Self` now?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 186 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQVMcij4OtTLnxKjDgz:-MQlob5U2IwPSsZfkyak:b2aylve) ([raw file](https://github.com/cockroachdb/cockroach/blob/d063a05a39929fc69bbdc04659b8daaf1ece9796/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L186)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It abbreviated some earlier name for `contentionEventHelper`. Fixed.
</blockquote></details>

Did you want to update the variable name?
___
*[pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self, line 87 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MRR3s_l-6h98h4Wn-Ws:-MRR3s_l-6h98h4Wn-Wt:b-aj5sq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/756037b892f1d2ef18b74b2435459ff76b844b47/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self#L87)):*
> ```
> [4] sequence reqTxn2: blocked on select in concurrency.(*lockTableWaiterImpl).WaitOn
> 
> # TODO(tbg): show state
> ```

Did you mean to address this TODO?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Dismissed @irfansharif from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58444)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 694 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQHBPxciWIj3amLRMzD-r14-694:-MRUFHfjC1WDQ3EB4Dgq:bpas5wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5265ba8c8818ee8cd465f29102533bc229ca37d4/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L694)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

It would be a bit easier to read the implementation below if we had an explicit early return for this case.
</blockquote></details>

Looking at this again, I agree. I also realized the code was buggy! It was resetting `tBegin` too frequently. The new implementation comes with a regression test for that, and the datadriven output didn't change (since the duration is mocked out there for determinism).
___
*[pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self, line 87 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MRR3s_l-6h98h4Wn-Ws:-MRUFcTi-az0pjH1PKtZ:b5bnudm) ([raw file](https://github.com/cockroachdb/cockroach/blob/756037b892f1d2ef18b74b2435459ff76b844b47/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self#L87)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you mean to address this TODO?
</blockquote></details>

I did leave a TODO for you in the block below. It looked nontrivial to expose this properly.
___
*[pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self, line 105 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQqen4OPUJvT0Dzm0jY-r15-105:-MRUFhgS2SuTfO_dpbsB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/756037b892f1d2ef18b74b2435459ff76b844b47/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self#L105)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Stray "doesn't"
</blockquote></details>

Done.
___
*[pkg/roachpb/api.proto, line 2211 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/58444#-MQHBPxciWIj3amLRMzO-r14-2211:-MRUFuWy74PVkk6DHnce:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5265ba8c8818ee8cd465f29102533bc229ca37d4/pkg/roachpb/api.proto#L2211)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/Txn/TxnMeta
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=irfansharif,nvanbenschoten

tbg commented:
bors r=irfansharif,nvanbenschoten
